contracts/eosio.bios/src/eosio.bios.cpp

```diff
@@ -54,4 +56,122 @@ void bios::reqactivated( const eosio::checksum256& feature_digest ) {
+
+    resource_config_table resource_config_singleton(get_self(), get_self().value);
+    resource_config config;
+
+    if (resource_config_singleton.exists()) {
+        // Singleton exists, get the existing config and modify the two values
+        config = resource_config_singleton.get();
+        config.ram_price = ram_price;
+        config.total_ram_available = total_ram_available;
+    } else {
+        // Singleton does not exist, set the two values and also set other values to 0
+        config = resource_config{ram_price, total_ram_available, 0, 0, 0};
+    }
+
+    // Save the modified or new config back to the singleton
+    resource_config_singleton.set(config, get_self());
 }
+
+void bios::buyram(eosio::name dao_owner, eosio::name app, eosio::asset quant) {
+    require_auth(app); // Check that the app has the necessary authorization
+
+   // Access the account table from id.tmy.hpp
+   idtmy::id::account_type_table account_type("id.tmy"_n, "id.tmy"_n.value);
+   // Check the account type of the app
+   auto itr = account_type.find(app.value);
+   eosio::check(itr != account_type.end(), "Could not find account");
+   eosio::check(itr->account_type == idtmy::enum_account_type::App, "Only apps can buy and sell RAM");
+
+   // Check that the RAM is being purchased with the correct token
+   eosio::check(quant.symbol == bios::system_resource_currency, "must buy ram with core token");
+
+   // Check that the amount of tokens being used for the purchase is positive
+   eosio::check(quant.amount > 0, "Amount must be positive");
+
+    // Check the decimal place of the token amount
+    eosio::check(quant.symbol.precision() <= 4, "Token amount must have up to 4 decimal places");
+
+    // Get the RAM price
+   resource_config_table resource_config_singleton(get_self(), get_self().value);
+    resource_config default_config{0, 0, 0, 0, 0, 0 };
+    auto config = resource_config_singleton.get_or_create(get_self(), default_config);
+
+    // Read values from the table
+    double ram_price = config.ram_price;
+    uint64_t ram_fee = (1.0 + config.ram_fee);
+    uint64_t ram_purchase = (ram_price / ram_fee) * quant.amount;
+
+    eosio::check(config.total_ram_available >= config.total_ram_used + ram_purchase, "Not enough RAM available");
+
+    // modify the values and save them back to the table,
+    config.total_ram_used += ram_purchase;
+    resource_config_singleton.set(config, get_self());
+    
+    //Allocate the RAM
+    int64_t myRAM, myNET, myCPU;
+    eosio::get_resource_limits( app, myRAM, myNET, myCPU );
+    eosio::set_resource_limits(app, myRAM + ram_purchase, myNET, myNET);
+
+
+   // Transfer token and buy ram
+     eosio::action(permission_level{get_self(), "active"_n},
+               "onocoin.tmy"_n,
+               "transfer"_n,
+        std::make_tuple(dao_owner, "gov.tmy"_n, quant, std::string("buy ram")))
+            .send();
+}
+
+void bios::sellram(eosio::name dao_owner, eosio::name app, int64_t bytes) {
```

Comment:
`sellram(eosio::name dao_owner, eosio::name app, eosio::asset quant)`

both functions should be consistent, and specify the amount to buy based on assets to use, or bytes to purchase